echochar.c

```diff
@@ -45,22 +46,158 @@ void echochar(uint8_t c) {
+							case 0xC2: prtflush("\xE2\x94\x82"); break; // 0x2502, │
+							case 0xC3: prtflush("\xE2\x94\x81"); break; // 0x2501, ━
+							case 0xC9: prtflush("\xE2\x95\xAE"); break; // 0x256E, ╮
+							case 0xCA: prtflush("\xE2\x95\xB0"); break; // 0x2570, ╰
+							case 0xCB: prtflush("\xE2\x95\xAF"); break; // 0x256F, ╯
+							case 0xCD: prtflush("\xE2\x95\xB2"); break; // 0x2572, ╲
+							case 0xCE: prtflush("\xE2\x95\xB1"); break; // 0x2571, ╱
+							case 0xD1: prtflush("\xE2\x97\x8F"); break; // 0x25CF, ●
+							case 0xD3: prtflush("\xE2\x99\xA5"); break; // 0x2665, ♥
+							case 0xD5: prtflush("\xE2\x95\xAD"); break; // 0x256D, ╭
+							case 0xD6: prtflush("\xE2\x95\xB3"); break; // 0x2573, ╳
+							case 0xD7: prtflush("\xE2\x97\x8B"); break; // 0x25CB, ○
+							case 0xD8: prtflush("\xE2\x99\xA3"); break; // 0x2663, ♣
+							case 0xDA: prtflush("\xE2\x99\xA6"); break; // 0x2666, ♦
+							case 0xDB: prtflush("\xE2\x94\xBC"); break; // 0x253C, ┼
+							case 0xDD: prtflush("\xE2\x94\x82"); break; // 0x2502, │
+							case 0xDE: prtflush("\xCF\x80"); break; // 0x3C0, π
+							case 0xDF: prtflush("\xE2\x97\xA5"); break; // 0x25E5, ◥
+							case 0xE0: prtflush("\xC2\xA0"); break; // 0xA0,  
+							case 0xE1: prtflush("\xE2\x96\x8C"); break; // 0x258C, ▌
+							case 0xE2: prtflush("\xE2\x96\x84"); break; // 0x2584, ▄
+							case 0xE3: prtflush("\xE2\x96\x94"); break; // 0x2594, ▔
+							case 0xE4: prtflush("\xE2\x96\x81"); break; // 0x2581, ▁
+							case 0xE5: prtflush("\xE2\x96\x8F"); break; // 0x258F, ▏
+							case 0xE6: prtflush("\xE2\x96\x92"); break; // 0x2592, ▒
+							case 0xE7: prtflush("\xE2\x96\x95"); break; // 0x2595, ▕
+							case 0xE9: prtflush("\xE2\x97\xA4"); break; // 0x25E4, ◤
+							case 0xEB: prtflush("\xE2\x94\x9C"); break; // 0x251C, ├
+							case 0xED: prtflush("\xE2\x94\x94"); break; // 0x2514, └
+							case 0xEE: prtflush("\xE2\x94\x90"); break; // 0x2510, ┐
+							case 0xEF: prtflush("\xE2\x96\x82"); break; // 0x2582, ▂
+							case 0xF0: prtflush("\xE2\x94\x8C"); break; // 0x250C, ┌
+							case 0xF1: prtflush("\xE2\x94\xB4"); break; // 0x2534, ┴
+							case 0xF2: prtflush("\xE2\x94\xAC"); break; // 0x252C, ┬
+							case 0xF3: prtflush("\xE2\x94\xA4"); break; // 0x2524, ┤
+							case 0xF4: prtflush("\xE2\x96\x8E"); break; // 0x258E, ▎
+							case 0xF5: prtflush("\xE2\x96\x8D"); break; // 0x258D, ▍
+							case 0xF9: prtflush("\xE2\x96\x83"); break; // 0x2583, ▃
+							case 0xFD: prtflush("\xE2\x94\x98"); break; // 0x2518, ┘
+							case 0xFF: prtflush("\xCF\x80"); break; // 0x3C0, π
 							default: prtnumflush("%c", c);
 						}
 						break;
 					case 1:
 						switch (c) {
-							case 126: prtflush("▒"); break;
-							case 127: prtflush("?"); break;
-							case 0xDF: prtflush("?"); break;
-							case 255: prtflush("▒"); break;
-							default: prtnumflush("%c", c);
+							case 0x5C: prtflush("\xC2\xA3"); break; // 0xA3, £
```

Comment:
This would be much better if it were a table that maps 8 bit PETSCII to 16 bit Unicode numbers, and you do the UTF-8 encoding separately.

(The UTF-8 in the comments is fine, and useful.)

Reply:
I don't know exactly what is the advantage with a table over a switch. A switch is more flexible and easier to maintain. I can shrink the switches further since there are still repetitions in the PETSCII. An array would contain a lot of elements that would have some special value because there is no translation for all PETSCII codes. I could change the cases in the switch so that they don't contain strings in most cases:
From:
case 0x5E: prtflush("\xE2\x86\x91"); break; // 0x2191, ↑
To:
case 0x5E: prtuptflush(0x2191); break; // ↑ upt = Unicode point

I did use int utf8_encode(char *out, uint32_t utf) to convert from Unicode point to UTF-8. I thought it would make the emulator faster if I did the conversion outside the emulator.

Maybe there are PETSCII characters outside 16 bit Unicode. A switch can handle these characters better.